core/engine.go

```diff
@@ -157,20 +158,24 @@ func (e *Engine) startBackgroundProcesses( //nolint:funlen
 	globalCtx, runCtx context.Context, runResult <-chan error, runSubCancel func(),
 ) (wait func()) {
 	processes := new(sync.WaitGroup)
+	metricsCtx, metricsDone := context.WithCancel(context.Background())
```

Comment:
I don't understand why this can't be globalCtx ?

Reply:
Because we need to ensure that the collector goroutines finish after `e.processMetrics()` is done, otherwise no metrics would be processed by them (in the case of JSON, the output file would be empty). This fixes the 2nd issue from the description.

Reply:
okay, I got it!

To be honest ... this probably isn't enough as there is nothing guaranteeing that just because we pushed metrics to the channel they will be processed before we also tell them to stop ... But it's probably a lot harder for this race to happen ... I think. 

This seems like something that a more complete rewrite such as required by https://github.com/loadimpact/k6/issues/1423 would be better able to produce. 
Although to be honest ... we can probably lose some data at the end and still be fine, but ti will be better if we don't :D

Reply:
Yeah, you're right, this needs better synchronization for collectors to ensure it can never happen, but I didn't want to get into a major refactor for this. Although I'll probably have to for #1423 :smile: